### PR TITLE
Do HTTP request tracing based on tracelogenable flag

### PR DESCRIPTION
**Pull request by arunkumarchacko**

### CODE REVIEW

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java [hunk 1]
```diff
@@ -301,8 +303,10 @@ public GoogleCloudStorageImpl(
    * @param storage {@link Storage} to use for I/O.
    */
   @SuppressWarnings("CheckReturnValue")
+  @VisibleForTesting
   public GoogleCloudStorageImpl(GoogleCloudStorageOptions options, Storage storage) {
-    this(options, storage, /* downscopedAccessTokenFn= */ null);
+    this(options, storage, null, tryGetCredentialsFromStorage(storage));
+    warnIfTracingEnabled(storageOptions.isTraceLogEnabled());
```
Comment (abmodi) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:309]: We should clean it up in future because this constructor should never be used.

**Review by abmodi — APPROVED** [review]

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java [hunk 2]
```diff
@@ -872,6 +870,78 @@ public void deleteObjects(List<StorageResourceId> fullObjectNames) throws IOExce
     }
   }
 
+  private void createGrpcStubIfEnabled(GoogleCloudStorageOptions options, Credentials credentials) {
+    if (!this.storageOptions.isGrpcEnabled()) {
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:874]: We usually use `this.` prefix only in constructors or setters when there is a conflict between class fields and method parameters

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java [hunk 3]
```diff
@@ -872,6 +870,78 @@ public void deleteObjects(List<StorageResourceId> fullObjectNames) throws IOExce
...
+      return;
+    }
+
+    this.watchdog =
+        Watchdog.create(Duration.ofMillis(options.getGrpcMessageTimeoutCheckInterval()));
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:879]: This is a non-obvious side effect for create method, can we move `watchdog` inside the gRPC stub object, for example?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java [hunk 4]
```diff
@@ -872,6 +870,78 @@ public void deleteObjects(List<StorageResourceId> fullObjectNames) throws IOExce
...
+
+    Credentials theCredentials = credentials;
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:881]: Do we really need `the` prefix? Usually we just use `credentials`

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java [hunk 5]
```diff
@@ -872,6 +870,78 @@ public void deleteObjects(List<StorageResourceId> fullObjectNames) throws IOExce
...
+    if (theCredentials == null) {
+      checkArgument(
+          httpRequestInitializer instanceof RetryHttpInitializer,
+          "request initializer must be an instance of the RetryHttpInitializer class or credential information should be supplied when gRPC API enabled");
+      theCredentials = ((RetryHttpInitializer) httpRequestInitializer).getCredentials();
+    }
+
+    this.storageStubProvider =
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:889]: Probably it will be better if we return stub here and move watchdog in the stub

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java [hunk 6]
```diff
@@ -872,6 +870,78 @@ public void deleteObjects(List<StorageResourceId> fullObjectNames) throws IOExce
...
+        StorageStubProvider.newInstance(
+            this.storageOptions, this.backgroundTasksThreadPool, theCredentials);
```
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:891]: nit: remove `this.` prefixes.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java (hunk 1 above)
Comment (medb) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java:309]: It seems that logic in the constructors of this class becomes more and more complicated and hard to follow. I think that we should bite a bullet and refactor it by using Builder class (or possible factory methods, but this seems less optimal) to provide input parameters for this class initialization instead of maintaining multiple contradictory constructors.

**Review by medb — CHANGES_REQUESTED** [review]

**Review by medb — APPROVED** [review]
